networkx/algorithms/flow/dinitz_alg.py

```diff
@@ -173,35 +173,52 @@ def dinitz_impl(G, s, t, capacity, residual, cutoff):
 
     def breath_first_search():
         parents = {}
+        vertex_dist = {}
         queue = deque([s])
+        distance = deque([0])
         while queue:
             if t in parents:
                 break
             u = queue.popleft()
+            dist = distance.popleft()
             for v in R_succ[u]:
                 attr = R_succ[u][v]
-                if v not in parents and attr["capacity"] - attr["flow"] > 0:
-                    parents[v] = u
-                    queue.append(v)
+                if attr["capacity"] - attr["flow"] > 0:
+                    if v in parents:
+                        if vertex_dist[v] == dist + 1:
+                            parents[v].append(u)
+                    else:
+                        parents[v] = [u]
+                        vertex_dist[v] = dist + 1
+                        queue.append(v)
+                        distance.append(dist + 1)
         return parents
 
     def depth_first_search(parents):
         """Build a path using DFS starting from the sink"""
-        path = []
-        u = t
-        flow = INF
-        while u != s:
-            path.append(u)
-            v = parents[u]
-            flow = min(flow, R_pred[u][v]["capacity"] - R_pred[u][v]["flow"])
-            u = v
-        path.append(s)
-        # Augment the flow along the path found
-        if flow > 0:
+        total_flow = 0
+        while True:
+            path = []
+            u = t
+            flow = INF
+            while u != s:
+                path.append(u)
+                for v in parents[u]:
+                    if R_pred[u][v]["capacity"] - R_pred[u][v]["flow"] > 0:
+                        break
```

Comment:
Why do we "break"? Why don't we look at all the parents -- we might choose one that doesn't lead to the source.

Also, it will speed things up to store `R_pred[u]` and then `R_pred[u][v]` and you might as well store the difference `R_pred[u][v]["capacity"] - R_pred[u][v]["flow"]` so you don't have to recompute that too.

Reply:
This part is updated in the latest version. I would be pleased to know your further advice.